src/wp-includes/option.php

```diff
@@ -1413,64 +1395,37 @@ function get_network_option( $network_id, $option, $default = false ) {
 		return $pre;
 	}
 
-	// Prevent non-existent options from triggering multiple queries.
-	$notoptions_key = "$network_id:notoptions";
-	$notoptions     = wp_cache_get( $notoptions_key, 'site-options' );
-
-	if ( is_array( $notoptions ) && isset( $notoptions[ $option ] ) ) {
-
+	if ( ! is_multisite() ) {
 		/**
 		 * Filters a specific default network option.
 		 *
 		 * The dynamic portion of the hook name, `$option`, refers to the option name.
 		 *
+		 * @param mixed $default The value to return if the site option does not exist
+		 *                           in the database.
+		 * @param string $option Option name.
+		 * @param int $network_id ID of the network.
+		 *
 		 * @since 3.4.0
 		 * @since 4.4.0 The `$option` parameter was added.
 		 * @since 4.7.0 The `$network_id` parameter was added.
 		 *
-		 * @param mixed  $default    The value to return if the site option does not exist
-		 *                           in the database.
-		 * @param string $option     Option name.
-		 * @param int    $network_id ID of the network.
 		 */
-		return apply_filters( "default_site_option_{$option}", $default, $option, $network_id );
-	}
-
-	if ( ! is_multisite() ) {
-		/** This filter is documented in wp-includes/option.php */
-		$default = apply_filters( 'default_site_option_' . $option, $default, $option, $network_id );
+		$default = apply_filters( "default_site_option_{$option}", $default, $option, $network_id );
 		$value   = get_option( $option, $default );
 	} else {
-		$cache_key = "$network_id:$option";
-		$value     = wp_cache_get( $cache_key, 'site-options' );
-
-		if ( ! isset( $value ) || false === $value ) {
-			$row = $wpdb->get_row( $wpdb->prepare( "SELECT meta_value FROM $wpdb->sitemeta WHERE meta_key = %s AND site_id = %d", $option, $network_id ) );
-
-			// Has to be get_row() instead of get_var() because of funkiness with 0, false, null values.
-			if ( is_object( $row ) ) {
-				$value = $row->meta_value;
-				$value = maybe_unserialize( $value );
-				wp_cache_set( $cache_key, $value, 'site-options' );
-			} else {
-				if ( ! is_array( $notoptions ) ) {
-					$notoptions = array();
-				}
-
-				$notoptions[ $option ] = true;
-				wp_cache_set( $notoptions_key, $notoptions, 'site-options' );
+		$meta = get_metadata_raw( 'site', $network_id, $option );
+		if ( is_array( $meta ) && ! empty( $meta ) ) {
+			$value = array_shift( $meta );
+		} else {
+			/** This filter is documented in wp-includes/option.php */
+			$value = apply_filters( "default_site_option_{$option}", $default, $option, $network_id );
 
-				/** This filter is documented in wp-includes/option.php */
-				$value = apply_filters( 'default_site_option_' . $option, $default, $option, $network_id );
-			}
+			/** This action is documented in wp-includes/meta.php */
+			$value = apply_filters( 'default_site_metadata', $value, $network_id, $option, true, 'site' );
```

Comment:
I'm a bit concerned about calling this here unconditionally. `get_site_option()` allows passing a `$default`, and any developer that uses `default_site_option_{$option}` should be aware of that already, which can be a relevant factor for how to implement the filter callback. For example you could say something like "if the default is `false`, replace it with my custom default here", almost as a shortcut to WP core's `$passed_default` boolean on the regular `default_option_{$option}` filter. (I almost wonder whether we should add that to this filter too.)

Any way, I am diverting a bit: My main concern here is that users of the `default_{$meta_type}_metadata` filters do not consider what the initial parameter value is since there is no way to manually pass a default to metadata functions. So I am wary about developers accidentally overriding this.

This is not necessarily a blocker to committing this, but I would be more comfortable if we only ran this code when no default was manually passed. This way we avoid accidentally modifying a passed default. In addition, it would allow us to simplify this code by actually using the function `get_metadata_default()`.
```suggestion
			// If default was neither passed manually nor modified by the filter, fall back to using metadata default.
			if ( func_num_args() <= 2 && false === $value ) {
				$value = get_metadata_default( 'site', $network_id, $option, true );
			}
```

Reply:
I think if you are going to use the `default_site_metadata` filter, you have to take ownership that. As a developer, it you would have to read the current value and take account of that. 

I do really see where you are coming from here. I think it should be kept how it is and highlighted in the dev note.

Reply:
The sooner this lands the more feedback we can get.